BibBuddy/app/src/main/java/de/bibbuddy/BookFragment.java

```diff
@@ -203,80 +240,169 @@ public void onClick(DialogInterface dialog, int which) {
-        Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
-      requestStoragePermission();
+  ActivityResultLauncher<Intent> filePickerActivityResultLauncher = registerForActivityResult(
+      new ActivityResultContracts.StartActivityForResult(),
+      new ActivityResultCallback<ActivityResult>() {
+        @Override
+        public void onActivityResult(ActivityResult result) {
+          if (result.getResultCode() == Activity.RESULT_OK) {
+            Intent data = result.getData();
+
+            if (data != null) {
+
+              Uri uri = data.getData();
+              if (importBibTex.isBibFile(UriUtils.getFullUriPath(context, uri))) {
+
+                handleImport(uri);
+
+                Toast.makeText(context, getString(R.string.imported_file_name_is) + '\n'
+                    + UriUtils.getUriFileName(getActivity(), uri), Toast.LENGTH_LONG).show();
+
+              } else {
+                showDialogNonBibFile();
+              }
+            }
+          }
+        }
+      });
+
+  private void handleImport(Uri uri) {
+    String bibText = readBibFile(uri);
+
+    List<String> nonRedundantBibItems
+        = importBibTex.getNonRedundantBibItems(bibText);
+
+    Book book;
+    for (int i = 0; i < nonRedundantBibItems.size(); i++) {
+
+      if (nonRedundantBibItems.get(i).startsWith(BibTexKeys.BOOK_TAG)) {
+        importBibTex.parseBibItem(nonRedundantBibItems.get(i));
+        book = importBibTex.importBook();
+        addImportedBook(book, importBibTex.parseAuthorNames());
+        importBibTex.importBibNote(noteDao, book);
+      }
+
+    }
+
+  }
+
+  private void addImportedBook(Book book, List<Author> authorList) {
+
+    if (bibTagValue.get(BibTexKeys.ANNOTE).isEmpty()) {
+      bookModel.addBook(book, authorList);
+
     } else {
-      // if the user has already allowed access to device external storage
-      exportBibTex.createBibFile();
-      exportBibTex.writeBibFile(exportBibTex.getBibDataFromShelf(shelfId, bookDao, noteDao));
-
-      Toast.makeText(getContext(),
-          getString(R.string.exported_file_stored_in) + '\n'
-              + File.separator + StorageKeys.DOWNLOAD_FOLDER + File.separator
-              + shelfName + StorageKeys.BIB_FILE_TYPE, Toast.LENGTH_LONG).show();
+      bookModel.addImportedBook(book, authorList);
     }
+
+    adapter.setBookList(bookModel.getCurrentBookList());
+    adapter.notifyDataSetChanged();
+    updateEmptyView(bookModel.getCurrentBookList());
   }
 
-  private void requestStoragePermission() {
-    if (ActivityCompat.shouldShowRequestPermissionRationale(getActivity(),
-        Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
+
+  private String readBibFile(Uri uri) {
+    try {
+      return importBibTex.readTextFromUri(uri);
+
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+
+    return null;
+  }
+
+  private void filePicker()  {
+    filePickerIntent = new Intent(Intent.ACTION_GET_CONTENT);
```

Comment:
From before the Intent should be a local variable.
```suggestion
    Intent filePickerIntent = new Intent(Intent.ACTION_GET_CONTENT);
```